### PR TITLE
#31: Fix bugs in WAL implementation due to incorrect handling of file offsets

### PR DESCRIPTION
**Pull request by melroy-tellis**
## Description

Resolve bugs arising from incorrect assumptions about the file offset in the WAL.
- Fix WriteEntry() and WriteDelete() to always seek to the end of the file.
- Update WriteEntry() to keep track of the offset in the file for the oldest pending entry.
- Fix Replay() to skip bytes for the key and value by advancing the file offset when encountering committed entries and return typed entries allowing callers to differentiate between pending and deleted entries.
- Fix MarkCommitted() to start from the offset of the oldest pending entry and continue marking all further entries as committed.

Fixes #31 

## Type of change

Please delete options that are not relevant.
- [✓] Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

Added additional tests `TestWAL/WriteEntryMarkCommitedWriteEntryReplay`, `TestWAL/ReplayAfterCommit2` and `TestWAL/MarkCommitted2` which ensures that the WAL handles writing of a new entry after a committed entry correctly without clobbering the previous entry. It also validates that Replay() skips committed records correctly.

Updated `TestWAL/WriteDelete` to expect a single entry with a Deleted flag instead of no entries, it was previously passing because of a bug in Replay() which was not consuming all the bytes for committed records causing the reading of subsequent entries from incorrect offsets in the file.

- [✓] TestWAL/WriteEntryMarkCommitedWriteEntryReplay
- [✓] TestWAL/MarkCommitted2
- [✓] TestWAL/ReplayAfterCommit2
- [✓] TestWAL/WriteDelete

## Checklist:

- [✓] My code follows the style guidelines of this project
- [✓] I have performed a self-review of my own code
- [✓] I have commented my code, particularly in hard-to-understand areas
- [✓] I have made corresponding changes to the documentation
- [✓] My changes generate no new warnings
- [✓] I have added tests that prove my fix is effective or that my feature works
- [✓] New and existing unit tests pass locally with my changes
- [✓] Any dependent changes have been merged and published in downstream modules

## Additional Notes
I see additional opportunities for refactoring the WAL implementation in line with DRY principles by refactoring WriteEntry and WriteDelete to use a common helper since the logic for both these operations is essentially the same (apart from tracking the last pending offset). I have already split out the logic for reading the header for an entry for sharing between Replay() and MarkCommitted()

### CODE REVIEW

File: internal/wal/wal.go [hunk 1]
```diff
@@ -33,33 +51,60 @@ func (w *WAL) WriteEntry(key, value string) error {
...
 
 	buf := make([]byte, 9+len(keyBytes))
 	binary.LittleEndian.PutUint32(buf[0:4], keySize)
 	binary.LittleEndian.PutUint32(buf[4:8], 0) // value size 0
-	buf[8] = 'D'                               // 'D' means delete
+	buf[8] = EntryDeleted
 	copy(buf[9:], keyBytes)
 
-	_, err := w.file.Write(buf)
+	_, err = w.file.Write(buf)
```
Comment (prabhat3697) [inline internal/wal/wal.go:107]: oldestPendingOffset is not updated for WriteDelete. This is used for key value storage

File: internal/storage/flat_meta_vector_storage.go [hunk 2]
```diff
@@ -714,12 +714,12 @@ func (e *FlatMetaVectorEngine) replayWAL() error {
...
 	}
 	for _, entry := range records {
-		keyBytes := []byte(entry[0])
+		keyBytes := []byte(entry.Key)
 		if len(keyBytes) != 8 {
 			return fmt.Errorf("invalid WAL key length: expected 8, got %d", len(keyBytes))
 		}
 		id := int64(binary.LittleEndian.Uint64(keyBytes))
-		if len(entry[1]) == 0 {
+		if entry.Flag == wal.EntryDeleted {
```
Comment (prabhat3697) [inline internal/storage/flat_meta_vector_storage.go:722]: @melroy-tellis Both "vector_storage" and "flat_meta_vector_storage" push empty value for a delete entry. With new wal implementation vector deletes will fall in "P" state which will try to replay it as empty vector. We need to fix the vector deletes as well to use EntryDeleted instead of empty values.

File: internal/wal/wal.go [hunk 3]
```diff
@@ -68,48 +113,68 @@ func (w *WAL) WriteDelete(key string) error {
...
+	offset := w.oldestPendingOffset
+	for offset >= 0 {
+		_, err := w.file.Seek(offset, io.SeekStart)
+		if err != nil {
+			return err
+		}
+		keySize, valSize, _, err := w.readHeader()
+		if err == io.EOF {
+			break
+		}
```
Comment (prabhat3697) [inline internal/wal/wal.go:142]: let's also add 

`if err != nil {
      return err
}
`

to handle any corrupted/truncated issues within wal.
In case of this unexpected error we can add warning logs as well.

**Review by prabhat3697 — COMMENTED** [review]
Overall solid implementation. Added comments on a few edge cases.
We can also target to add more tests like below if possible:

  • KV Delete(key), then Put(key, newValue), restart/replay, expect newValue.
  • Pending delete replay for vector and flat-meta engines.
  • WriteDelete() followed by MarkCommitted() should not leave a replayable delete.
  • Truncated/Corrupted WAL header behavior in MarkCommitted().

File: internal/wal/wal.go [hunk 4]
```diff
@@ -8,21 +8,39 @@ import (
...
 
+type WALEntry struct {
+	Key   string
+	Value string
+	Flag  byte
+}
+
+const (
+	EntryPending  byte = 'P'
+	EntryCommited byte = 'C'
```
Comment (prabhat3697) [inline internal/wal/wal.go]: TYPO:
EntryCommitted

**Review by prabhat3697 — APPROVED** [review]
Looks Good
